File: src/main/java/dev/todaka/robustredis/cluster/ClusterRouter.java
package dev.todaka.robustredis.cluster;

import dev.todaka.robustredis.NodeConnection;
import dev.todaka.robustredis.RedisRouter;
import dev.todaka.robustredis.connection.RedisURI;
import dev.todaka.robustredis.protocol.RedisCommand;

import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.CompletableFuture;

public class ClusterRouter implements RedisRouter {
    /**
     * SlotId to RedisURI mapping built by CLUSTER NODES response.
     * Contains exactly 16,384 entries.
     */
    private List<RedisURI> slotIdToNode;

    /**
     * RedisURI to NodeConnection mapping.
     */
    private final Map<RedisURI, CompletableFuture<NodeConnection>> connectionPool = new HashMap<>();

    public ClusterRouter(List<ClusterNodeView> views) {
        final var newSlotIdToNode = new ArrayList<RedisURI>(Collections.nCopies(16384, null));
        for (final var view : views) {
            final var uri = new RedisURI(view.getIp(), view.getPort());
            for (final var slotRange : view.getSlots()) {
                for (int s = slotRange.getFrom(); s <= slotRange.getTo(); s++) {
                    newSlotIdToNode.set(s, uri);
                }
            }
        }
        slotIdToNode = newSlotIdToNode;
    }

    public CompletableFuture<NodeConnection> findOrEstablishConnection(RedisCommand<?> command) {
        int slot;
        if (command.getFirstKey().isEmpty()) {
            // use slot = 0 for command without key
            slot = 0;
        } else {
            slot = CRC16.crc16(command.getFirstKey().getBytes(StandardCharsets.UTF_8)) % 16384;
        }
        final RedisURI uri = slotIdToNode.get(slot);

        synchronized (connectionPool) {
            return connectionPool.computeIfAbsent(
                    uri, notUsed -> NodeConnection.connectAsync(uri));
        }
    }
}

File: src/main/kotlin/dev/todaka/robustredis/cluster/ClusterRouter.kt
package dev.todaka.robustredis.cluster

import dev.todaka.robustredis.cluster.CRC16.crc16
import dev.todaka.robustredis.connection.NodeConnection
import dev.todaka.robustredis.connection.RedisURI
import dev.todaka.robustredis.exception.RedisInvalidClusterException
import dev.todaka.robustredis.model.RedisCommand
import java.util.concurrent.CompletableFuture

class ClusterRouter(views: List<ClusterNodeView>) {
    /**
     * SlotId to RedisURI mapping built by CLUSTER NODES response.
     * Contains exactly 16,384 entries.
     */
    private val slotIdToNode: List<RedisURI>

    /**
     * RedisURI to NodeConnection mapping.
     */
    private val connectionPool: MutableMap<RedisURI, CompletableFuture<NodeConnection>> = HashMap()

    init {
        val uris = views.flatMap { view ->
            view.slots.flatMap { slot ->
                (slot.from..slot.to).map { slotId -> slotId to RedisURI(view.ip, view.port) }
            }
        }
        if (uris.size != 16384) {
            throw RedisInvalidClusterException("Invalid CLUSTER NODES response")
        }

        slotIdToNode = uris.sortedBy { it.first }.map { it.second }
    }

    fun findOrEstablishConnection(command: RedisCommand<*>): CompletableFuture<NodeConnection> {
        val slot = command.commandInput.keys?.firstOrNull()
            ?.let { crc16(it.toByteArray(Charsets.UTF_8)) % 16384 }
            ?: 0 // use slot = 0 for command without key

        val uri = slotIdToNode[slot]
        synchronized(connectionPool) {
            return connectionPool.computeIfAbsent(uri) {
                NodeConnection.connectAsync(uri)
            }
        }
    }
}
